fix collision

psychocoderHPC opened the pull request:
With #4685 we missed to rewrite collision to support a frame size independent of the number of cells in the supercell.

This fix rewrites the required code to support any frame size and optimizes the particle access via the particle index.
This PR is optimizing the performance by a factor of ~20.

@pordyna Could you please validate the results? The README is pointing to SMILEI results which are not available, therefore a validation from my side is not possible.

A minimal fix was not possible therefore a larger refactoring was required.

# Review:
Check the code changes without white space changes! https://github.com/ComputationalRadiationPhysics/picongpu/pull/4745/files?w=1

psychocoderHPC committed:
fix collision

With #4685 we missed to rewrite collision to support a frame size independent of the number of cells in the supercell.

This fix rewrites the required code to support any frame size and optimizes the particle access via the particle index.

psychocoderHPC commented on include/picongpu/particles/collision/InterCollision.hpp:
```diff
@@ -29,480 +29,427 @@
+                {
+                    auto par0 = parAccessor0[i % size0];
+                    auto par1 = parAccessor1[i % size1];
+                    destCollisionFunctor.duplicationCorrection
```
@pordyna We need to talk about this duplication correction to take care that I have not broken anything.

pordyna replied:
well this does look wrong, but in the test case that I posted the sizes are equal, and the duplication correction is 1 so it is not the main problem

psychocoderHPC commented on include/picongpu/particles/collision/IntraCollision.hpp:
```diff
@@ -29,375 +29,355 @@
-                    T_TimesCollidedBox timesCollidedBox) const
-                {
-                    using namespace pmacc::particles::operations;
+            using Frame = typename T_ParBox::FrameType;
```
can be removed

pordyna commented:
I'm running tests now

pordyna commented:
@psychocoderHPC  So with the fix it is much faster but also much broken^^ We need to take a look at the changes.

## previous results:
![main_0](https://github.com/ComputationalRadiationPhysics/picongpu/assets/41141557/150a3a14-269b-40b1-b4d9-46a04f98e9f0)
![debug_0](https://github.com/ComputationalRadiationPhysics/picongpu/assets/41141557/bfb89dad-c318-43f9-83c9-47ca50b66456)
## new results:
![main_0](https://github.com/ComputationalRadiationPhysics/picongpu/assets/41141557/00faf88a-cf39-4e9e-8c69-84957f395905)
![debug_0](https://github.com/ComputationalRadiationPhysics/picongpu/assets/41141557/3a79105a-4edd-4c32-b425-d2251bda0ed3)

pordyna commented:
btw this is with 500ppc not 5000 as the plot says

pordyna commented on include/picongpu/particles/collision/InterCollision.hpp:
```diff
@@ -29,480 +29,425 @@
-                            destCollisionFunctor(detail::makeCollisionContext(worker, rngHandle), parLong, parShort);
-                        }
+            auto destCollisionFunctor
+                = srcCollisionFunctor(worker, localSuperCellOffset, densityLongList, densityShortList, maxListLength);
```
```suggestion
                = srcCollisionFunctor(worker, localSuperCellOffset, density0, density1, maxListLength);
```

pordyna commented on include/picongpu/particles/collision/InterCollision.hpp:
```diff
@@ -29,480 +29,425 @@
+            uint32_t const minListLength = math::min(size0, size1);
+            uint32_t const maxListLength = math::max(size0, size1);
+            float_X const densityLongList = maxListLength == size0 ? density0 : density1;
+            float_X const densityShortList = densityLongList == density1 ? density0 : density1;
```
```suggestion
```

pordyna commented on include/picongpu/particles/collision/InterCollision.hpp:
```diff
@@ -29,480 +29,425 @@
+                    auto par0 = parAccessor0[i % size0];
+                    auto par1 = parAccessor1[i % size1];
+                    destCollisionFunctor(detail::makeCollisionContext(worker, rngHandle), par0, par1);
+                }
```
```suggestion
                for(uint32_t i = numDupParticles; i < maxListLength; ++i)
                {
                    destCollisionFunctor.duplicationCorrection
                        = duplicationCorrection(i, minListLength, maxListLength);
                    auto par0 = parAccessor0[i % size0];
                    auto par1 = parAccessor1[i % size1];
                    destCollisionFunctor(detail::makeCollisionContext(worker, rngHandle), par0, par1);
                }
```

This was missing. Also there is no need to set it to 1, we are changing it anyway.

pordyna commented on include/picongpu/particles/collision/detail/ListEntry.hpp:
```diff
@@ -28,273 +28,368 @@
+             */
+            DINLINE auto operator[](uint32_t idx) const
+            {
+                return m_framePtrList[idx / frameSize][idx % frameSize];
```
```suggestion
                const uint32_t inSuperCellIdx = m_parIdxList[idx];
                return m_framePtrList[inSuperCellIdx / frameSize][inSuperCellIdx % frameSize];
```
If i understand ithe code correctly, this is most likely the fix to our problem.  I will check if it helps.

pordyna replied:
With this fix it looks good again, but I still need to rerun all tests in full.

psychocoderHPC replied:
nice catch!

pordyna commented:
@psychocoderHPC I looked at all test cases and compared visually to the results in our docs, it looks good to me!

psychocoderHPC commented:
I applied @pordyna fixes

psychocoderHPC commented on include/picongpu/particles/collision/detail/ListEntry.hpp:
```diff
@@ -28,273 +28,369 @@
+                        numParticles[cellIdx] = 0u;
+                        uint32_t numParsInCell = numParArray[cellIdx];
+
+                        constexpr uint32_t chunkSize = cellListChunkSize * frameSize * sizeof(uint32_t);
```
offline review with @pordyna : **BUG**  `frameSize` must be removed because we need the number of particles per cell and not per frame

psychocoderHPC replied:
This is now fixed too

psychocoderHPC commented:
@pordyna can you review the PR and give a green thump if you are ok with the changes

pordyna commented:
@psychocoderHPC were there any significant changes since the last status?

psychocoderHPC commented:
> @psychocoderHPC were there any significant changes since the last status?

No I only removed unused variables, all other is on the level we assembled together.

psychocoderHPC commented:
I tested the beam collision example case 0 with 512 frame slots and the results look equal to the default param set with 256 as frame size.

![BR_main_0](https://github.com/ComputationalRadiationPhysics/picongpu/assets/4037125/2fb2b4ef-894d-472f-a0c5-061f00f08364)

![BR_debug_0](https://github.com/ComputationalRadiationPhysics/picongpu/assets/4037125/d089a5eb-8688-4a20-80fe-e53fb6e49668)

pordyna commented:
Looks good to me!

steindev commented on include/picongpu/particles/collision/detail/ListEntry.hpp:
```diff
@@ -28,273 +28,386 @@
+                        uint32_t numFrames = (numParticlesInSupercell + frameSize - 1u) / frameSize;
+                        constexpr uint32_t framePtrBytes = sizeof(typename T_ParticlesBox::FramePtr);
+
+                        // Chunk size is the typical number of frames based on cellListChunkSize which the number of
```
There is something wrong in this sentence.

psychocoderHPC replied:
added a variable and added a better description

steindev commented on include/picongpu/particles/collision/detail/ListEntry.hpp:
```diff
@@ -28,273 +28,386 @@
+
+            /* Update cell particle index list
+             *
+             * Update the contiguous index list per cell with particle indies.
```
typo indices

steindev commented on include/picongpu/particles/collision/detail/ListEntry.hpp:
```diff
@@ -28,273 +28,386 @@
-                    T_Filter& filter)
-                {
-                    constexpr uint32_t frameSize = T_ParBox::frameSize;
+            /** Creates an access to access particles within a cell
```
Creates an accessor?

steindev commented on include/picongpu/particles/collision/detail/ListEntry.hpp:
```diff
@@ -28,273 +28,386 @@
-                    uint32_t const numParticlesInSupercell,
-                    T_EntryListArray& parCellList,
-                    T_Filter& filter)
+        //! Cound number of particles per cell in the supercell
```
typo count

steindev commented on include/picongpu/particles/collision/detail/ListEntry.hpp:
```diff
@@ -28,273 +28,386 @@
+        /* Swap two list entries.
+         *
+         * @param v0 index of the 1st entry to swap
+         * @param v0 index of the 2nd entry to swap
```
typo `@param v1`

steindev commented on include/picongpu/particles/collision/detail/ListEntry.hpp:
```diff
@@ -28,273 +28,386 @@
+         *
+         * Shuffle the given array with indicies.
+         *
+         * @param worker lockstep worker
```
twice 'workeer'?

pordyna replied:
1. is the parameter name the other is part of the description

steindev commented on include/picongpu/particles/collision/detail/ListEntry.hpp:
```diff
@@ -28,273 +28,386 @@
+    namespace detail
     {
-        namespace collision
+        /** Index access to particles within a cell of an supercell.
```
an -> a

But I don't understand what you mean by 'Index access'

psychocoderHPC replied:
changed it to `Access all particles in a cell.` The index is explains in the doxygen of the `operator[]`

steindev commented on include/picongpu/particles/collision/IntraCollision.hpp:
```diff
@@ -29,375 +29,350 @@
+                            // intraCollisions (compare with 29 and use m_aa = 1/2 m_a). Bu this seams to be a
+                            // typo smilei doesn't include this extra factor. It was applied here in the
+                            // previous version, via the duplication correction.
+
```
How do you know which one is correct?

pordyna replied:
I'm quite sure that I followed the derivation in the paper and came to the conclusion that this factor was an error.

steindev reviewed (changes requested):
Few questions. And typos. But I cannot really do a review here, I cannot understand the details of this (and I probably should read the documentation to at least get the big picture.... :sweat_smile: )

I trust on @pordyna that he knows what is going on and that the tests validate  the results.

psychocoderHPC commented:
@steindev I fixed the typos and changed some of the documentation issues. Forced pushed it to the PR.
